pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java

```diff
@@ -218,24 +205,20 @@ public boolean readCompacted() {
      *
      * @return a SendMessageInfo object that contains the detail of what was sent to consumer
      */
-
-    public ChannelPromise sendMessages(final List<Entry> entries, EntryBatchSizes batchSizes, EntryBatchIndexesAcks batchIndexesAcks,
+    public Future<Void> sendMessages(final List<Entry> entries, EntryBatchSizes batchSizes, EntryBatchIndexesAcks batchIndexesAcks,
                int totalMessages, long totalBytes, long totalChunkedMessages, RedeliveryTracker redeliveryTracker) {
         this.lastConsumedTimestamp = System.currentTimeMillis();
-        final ChannelHandlerContext ctx = cnx.ctx();
-        final ChannelPromise writePromise = ctx.newPromise();
 
         if (entries.isEmpty() || totalMessages == 0) {
             if (log.isDebugEnabled()) {
                 log.debug("[{}-{}] List of messages is empty, triggering write future immediately for consumerId {}",
                         topicName, subscription, consumerId);
             }
-            writePromise.setSuccess();
             batchSizes.recyle();
             if (batchIndexesAcks != null) {
                 batchIndexesAcks.recycle();
             }
-            return writePromise;
+            return ImmediateEventExecutor.INSTANCE.newSucceededFuture(null);
```

Comment:
What is the main motivation for this change?

Reply:
It's because protocol handlers may not have a `ChannelHandlerContext` to create a `ChannelPromise` from (eg. in gRPC you don't have direct access to the Netty channel).

Reply:
I can move this part of code to the cnx delegate if you prefer (so there won't be any change for the regular ServerCnx/PulsarCommandSenderImpl).

Reply:
The code below is really tied to the Consumer so it cannot be moved to the PulsarCommandSender.

Reply:
I added a `newPromise` method to `TransportCnx` which `ServerCnx` implements with `ctx.newPromise()` so there's no change in how it works for the regular `ServerCnx`.